masonry/src/text/editor.rs

```diff
@@ -0,0 +1,672 @@
+        self.editor.set_selection(self.editor.selection.move_lines(
+            &self.editor.layout,
+            isize::MIN,
+            true,
+        ));
+    }
+
+    /// Move the selection focus point to the start of the physical line.
+    pub fn select_to_line_start(&mut self) {
+        self.editor
+            .set_selection(self.editor.selection.line_start(&self.editor.layout, true));
+    }
+
+    /// Move the selection focus point to the end of the buffer.
+    pub fn select_to_text_end(&mut self) {
+        self.editor.set_selection(self.editor.selection.move_lines(
+            &self.editor.layout,
+            isize::MAX,
+            true,
+        ));
+    }
+
+    /// Move the selection focus point to the end of the physical line.
+    pub fn select_to_line_end(&mut self) {
+        self.editor
+            .set_selection(self.editor.selection.line_end(&self.editor.layout, true));
+    }
+
+    /// Move the selection focus point up to the nearest cluster boundary on the previous line, preserving the horizontal position for repeated movements.
+    pub fn select_up(&mut self) {
+        self.editor.set_selection(
+            self.editor
+                .selection
+                .previous_line(&self.editor.layout, true),
+        );
+    }
+
+    /// Move the selection focus point down to the nearest cluster boundary on the next line, preserving the horizontal position for repeated movements.
+    pub fn select_down(&mut self) {
+        self.editor
+            .set_selection(self.editor.selection.next_line(&self.editor.layout, true));
+    }
+
+    /// Move the selection focus point to the next cluster left in visual order.
+    pub fn select_left(&mut self) {
+        self.editor
+            .set_selection(self.editor.selection.previous_visual(
+                &self.editor.layout,
+                self.editor.cursor_mode,
+                true,
+            ));
+    }
+
+    /// Move the selection focus point to the next cluster right in visual order.
+    pub fn select_right(&mut self) {
+        self.editor.set_selection(self.editor.selection.next_visual(
+            &self.editor.layout,
+            self.editor.cursor_mode,
+            true,
+        ));
+    }
+
+    /// Move the selection focus point to the next word boundary left.
+    pub fn select_word_left(&mut self) {
+        self.editor.set_selection(
+            self.editor
+                .selection
+                .previous_word(&self.editor.layout, true),
+        );
+    }
+
+    /// Move the selection focus point to the next word boundary right.
+    pub fn select_word_right(&mut self) {
+        self.editor
+            .set_selection(self.editor.selection.next_word(&self.editor.layout, true));
+    }
+
+    /// Select the word at the point.
+    pub fn select_word_at_point(&mut self, x: f32, y: f32) {
+        self.refresh_layout();
+        self.editor
+            .set_selection(Selection::word_from_point(&self.editor.layout, x, y));
```

Comment:
The behavior of this doesn't seem correct when I use the editor. It selects the work then something messes it up so that it only selects the word left of the pointer. I don't know if the problem is here, where this function is called, or in Parley.
If this isn't addressed in this PR it would make sense to file an issue to fix it later. This isn't a blocker because there was no double click to select in the prior code.

Reply:
This should reproduce in Parley's `vello_editor` example, I think